libraries/Microsoft.Bot.Builder.Dialogs/DialogSet.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    /// <summary>
+    /// A related set of dialogs that can all call each other.
+    /// </summary>
+    /// <typeparam name="C">TurnContext</typeparam>
+    public class DialogSet
+    {
+        private IDictionary<string, Dialog> _dialogs;
+
+        public DialogSet()
+        {
+            _dialogs = new Dictionary<string, Dialog>();
+        }
+
+        /// <summary>
+        /// Adds a new dialog to the set and returns the added dialog.
+        /// </summary>
+        public Dialog Add(string dialogId, Dialog dialog)
+        {
+            if (_dialogs.ContainsKey(dialogId))
+            {
+                throw new Exception($"DialogSet.add(): A dialog with an id of '{dialogId}' already added.");
+            }
+            return _dialogs[dialogId] = dialog;
+        }
+
+        /// <summary>
+        /// Adds a new waterfall to the set and returns the added waterfall.
+        /// </summary>
+        public Waterfall Add(string dialogId, WaterfallStep[] steps)
+        {
+            var waterfall = new Waterfall(steps);
+            Add(dialogId, waterfall);
+            return waterfall;
+        }
+
+        public DialogContext CreateContext(ITurnContext context, object state)
```

Comment:
>object state [](start = 65, length = 12)

This is one area where I think Tom and Chris should weigh in.  In the JavaScript version I have the developer just pass in empty JSON object { } which I will then populate with the stack.  The idea is the caller should have to care about what I need to persist state wise so I just want a bag to write to.  I don't know how well that works in C#

Reply:
If this is the semantic then it should be Dictionary<string, object> 

---
In reply to: [181199494](https://github.com/Microsoft/botbuilder-dotnet/pull/408#discussion_r181199494) [](ancestors = 181199494)

Reply:
Will that serialize ok as part of the bots overall state?  To be concrete the goal was to let the developer do something like this:

```JS
conversationState.myState = 'foo';
conversationState.dialogState = {};
const dc = dialogs.createContext(context, conversationState.dialogState);
```

Notice that the dialog sets state is completely isolated from the bots other state and to reset the dialogs state all up you can simply `delete conversationState.dialogState`.  This also makes it super simple to save off the state of a task while you do something else and then return to it at some point in the future.

Reply:
Internally it is Dictionary<string, object> where ultimately object must be serializable to JSON. (So you could have JObject as the alternative here.)

Reply:
I would suggest _not_ surfacing `JObject`. That's an underlying serialization implementation detail, so to expose it would be a leaky abstraction. 

It should be likely strongly typed via a generic parameter IMHO. I haven't looked at this in any depth yet, but I'd venture a guess there would be a fair amount of work to get to that right now. So, at the very least, I would agree with going for `Dictionary<K,V>` initially.